src/components/ColumnPicker/useColumns.ts

```diff
@@ -0,0 +1,102 @@
+import useStateFromProps from "@dashboard/hooks/useStateFromProps";
+import { addAtIndex, removeAtIndex } from "@dashboard/utils/lists";
+import { GridColumn } from "@glideapps/glide-data-grid";
+import React from "react";
+
+import { AvailableColumn } from "../Datagrid/types";
+
+export interface ColumnCategory {
+  name: string;
+  prefix: string;
+  availableNodes: AvailableColumn[];
+  selectedNodes: AvailableColumn[];
+  onSearch?: (query: string) => void;
+  hasNextPage: boolean;
+  hasPreviousPage: boolean;
+  onNextPage: (query: string) => void;
+  onPreviousPage: (query: string) => void;
+}
+
+export interface UseColumnsProps {
+  staticColumns: AvailableColumn[];
+  columnCategories?: ColumnCategory[];
+  selectedColumns: string[];
+  onSave: (columns: string[]) => void;
+  columnPickerSettings: string[];
+  setDynamicColumnSettings: (cols: string[]) => void;
+}
+
+export const useColumns = ({
+  staticColumns,
+  selectedColumns,
+  columnCategories,
+  onSave,
+  columnPickerSettings,
+  setDynamicColumnSettings,
+}: UseColumnsProps) => {
+  const dynamicColumns: AvailableColumn[] = React.useMemo(
+    () =>
+      columnCategories.reduce((columns: AvailableColumn[], category) => {
+        category.selectedNodes.forEach(column => {
+          if (columnPickerSettings.includes(column.id)) {
+            columns.push(column);
+          }
+        });
+        return columns;
+      }, [] as AvailableColumn[]),
+    [columnCategories, columnPickerSettings],
+  );
+
+  const initialColumnsState = React.useMemo(
+    () =>
+      [...staticColumns, ...dynamicColumns].filter(
+        column => selectedColumns.includes(column.id) || column.id === "empty",
+      ),
+    [dynamicColumns, staticColumns, selectedColumns],
+  );
+
+  const [visibleColumns, setVisibleColumns] =
+    useStateFromProps(initialColumnsState);
+
+  const onMove = React.useCallback(
+    (startIndex: number, endIndex: number): void => {
+      setVisibleColumns(old =>
+        addAtIndex(old[startIndex], removeAtIndex(old, startIndex), endIndex),
+      );
+    },
+    [setVisibleColumns],
```

Comment:
question: should you handle empty column here, like don't allow to move normal column behind empty?

Reply:
Great catch.

Reply:
Still don't see empty column handling

Reply:
Fixed now.